File: api/src/main/java/com/eirsteir/coffeewithme/api/exception/CWMException.java
package com.eirsteir.coffeewithme.api.exception;

import org.springframework.stereotype.Component;

@Component
public class CWMException {

    public static RuntimeException getException(EntityType entityType, ExceptionType exceptionType, String... args) {
        String messageTemplate = MessageTemplateUtil.getMessageTemplate(entityType, exceptionType);
        return getException(exceptionType, messageTemplate, args);
    }

    public static class EntityNotFoundException extends RuntimeException {
        public EntityNotFoundException(String message) {
            super(message);
        }
    }

    public static class DuplicateEntityException extends RuntimeException {
        public DuplicateEntityException(String message) {
            super(message);
        }
    }

    public static class InvalidStatusChangeException extends RuntimeException {
        public InvalidStatusChangeException(String message) {
            super(message);
        }
    }

    private static RuntimeException getException(ExceptionType exceptionType, String messageTemplate, String... args) {
        if (ExceptionType.ENTITY_NOT_FOUND.equals(exceptionType))
            return new EntityNotFoundException(MessageTemplateUtil.format(messageTemplate, args));
        else if (ExceptionType.DUPLICATE_ENTITY.equals(exceptionType))
            return new DuplicateEntityException(MessageTemplateUtil.format(messageTemplate, args));
        else if (ExceptionType.INVALID_STATUS_CHANGE.equals(exceptionType))
            return new InvalidStatusChangeException(MessageTemplateUtil.format(messageTemplate, args));

        return new RuntimeException(MessageTemplateUtil.format(messageTemplate, args));
    }

}

File: commons/src/main/kotlin/com/eirsteir/coffeewithme/commons/exception/CWMException.kt
package com.eirsteir.coffeewithme.commons.exception

import com.eirsteir.coffeewithme.commons.exception.MessageTemplateUtil
import org.springframework.stereotype.Component
import java.text.MessageFormat
import java.util.*


class CWMException(private val propertiesConfig: PropertiesConfig) {

    class EntityNotFoundException(message: String?) : RuntimeException(message)
    class DuplicateEntityException(message: String?) : RuntimeException(message)
    class InvalidStatusChangeException(message: String?) : RuntimeException(message)

    fun getException(
        entityType: EntityType, exceptionType: ExceptionType, vararg args: String
    ): RuntimeException {
        val messageTemplate: String = getMessageTemplate(entityType, exceptionType)
        return getException(exceptionType, messageTemplate, *args)
    }

    private fun getException(
        exceptionType: ExceptionType, messageTemplate: String, vararg args: String
    ) =
        when {
            ExceptionType.ENTITY_NOT_FOUND == exceptionType -> EntityNotFoundException(
                format(
                    messageTemplate,
                    *args
                )
            )
            ExceptionType.DUPLICATE_ENTITY == exceptionType -> DuplicateEntityException(
                format(
                    messageTemplate,
                    *args
                )
            )
            ExceptionType.INVALID_STATUS_CHANGE == exceptionType -> InvalidStatusChangeException(
                format(messageTemplate, *args)
            )
            else -> RuntimeException(format(messageTemplate, *args))
        }

    private fun getMessageTemplate(entityType: EntityType, exceptionType: ExceptionType): String {
        return entityType.name + "." + exceptionType.value.lowercase()
    }

    private fun format(template: String, vararg args: String): String {
        val templateContent = Optional.ofNullable(
            propertiesConfig.getConfigValue(template)
        )
        return templateContent
            .map { s: String -> MessageFormat.format(s, *args) }
            .orElseGet { String.format(template, *args) }
    }
}